proofs/alf/cvc5/programs/Strings.smt3

```diff
@@ -630,3 +682,158 @@
+                                                       (alf.match ((bb Bool))
+                                                          ($str_re_consume_rec s1 r1r @result.null rev)
+                                                          (
+                                                            (false  false)
+                                                            (bb     ($str_re_consume_rec s1 r2 ($result_combine bb b) rev))
+                                                          )
+                                                       )))
+    (($str_re_consume_rec s1 re.all @result.null rev)  (str.in_re "" @re.empty)) ; only used if re.all appears in unexpected position
+    (($str_re_consume_rec s1 re.all b rev)             b)                        ; end of re.inter
+    ; Union ignores conflicts, and otherwise combines the results. We report a conflict if all children give conflicts.
+    (($str_re_consume_rec s1 (re.union r1 r2) b rev)   (let ((r1r ($re_to_flat_form r1 rev)))
+                                                       (alf.match ((bb Bool))
+                                                          ($str_re_consume_rec s1 r1r @result.null rev)
+                                                          (
+                                                            (false  ($str_re_consume_rec s1 r2 b rev))
+                                                            (bb     ($str_re_consume_rec s1 r2 ($result_combine bb b) rev))
+                                                          )
+                                                       )))
+    (($str_re_consume_rec s1 re.none @result.null rev) false)     ; end of re.union, conflict, also used if re.none appears in unexpected position
+    (($str_re_consume_rec s1 re.none b rev)            b)         ; end of re.union, no conflict
+    (($str_re_consume_rec s1 r1 @result.null rev)      (str.in_re s1 r1))
+  )
+)
+
+; `$str_re_consume
+; args:
+; - s String: The string argument of the membership to rewrite.
+; - r RegLan: The regular expression argument of the membership to rewrite.
+; returns: >
+;   false if `(str.in_re s r)` can be shown  to be equivalent to false, or
+;   otherwise `(str.in_re sr rr)` where sr and rr are the result of "consuming"
+;   prefixes/suffixes from s and r. We consume from the end of s and r first by
+;   reversing them initially and converting them to flat forms. We then reverse
+;   them again and consume from the beginning of the remainder. Finally, we convert
+;   back from flat form when applicable.
+(define $str_re_consume ((s String) (r RegLan))
+  (let ((ss (string_to_flat_form s true)))  
+  (let ((rr ($re_to_flat_form r true)))
+  (alf.match ((s1 String) (r1 RegLan))
+      ($str_re_consume_rec ss rr @result.null true)
+      (
+        (false              false)
+        ((str.in_re s1 r1)  (let ((s1r ($str_rev true s1)) (r1r ($re_to_flat_form r1 true)))
```

Comment:
I'm undecided if we should have specific versions of reverse. The generic version of reverse is now possible (https://github.com/cvc5/cvc5/blob/main/proofs/alf/cvc5/programs/Nary.smt3#L63) to simplify so that it only takes one argument.  I will likely refactor to switch the string concat reverse to this later.